kbd-mode.el

```diff
@@ -151,6 +151,190 @@ form `@MACRO-NAME' with `kbd-mode-variable-name-face'."
+
+Use \\[kbd-mode-stop-demo] to stop current demo kmonad instance.
+If configuration file has errors, demo will not start
+instead, a compilation buffer will be shown."
+  (interactive)
+  (when (shell-command "ps -C kmonad")
+    (progn
+      (setq kbd-mode-had-kmonad t))
+    (kbd-mode--old-kmonad-suspend))
+  (setq kbd-mode-demo-file
+        (kbd-mode--find-kbd-file
+         (buffer-file-name (current-buffer))))
+  (kbd-mode--create-demo-buffer)
+  (pop-to-buffer "*kmonad-demo*")
+  (kbd-mode-demo-mode t))
+
+;;;###autoload
+(defun kbd-mode-stop-demo ()
+  "Stop the demo."
+  (interactive)
+  (with-current-buffer "*kmonad-demo*"
+    (kbd-mode-demo-mode 0)
+    (kill-buffer-and-window)
+    (kbd-mode--kill-kmonad-process)
+    (when kbd-mode-had-kmonad
+      (kbd-mode--old-kmonad-resume))))
+
+;;;###autoload
+(defun kbd-mode-goto-demo ()
+  "Goto demo window."
+  (interactive)
+  (select-window (get-buffer-window "*kmonad-demo*")))
+
+;;;###autoload
+(defun kbd-mode-goto-config ()
+  "Go to config file.
+Accessible only from demo buffer."
+  (interactive)
+  (when kbd-mode-demo-mode
+    (select-window
+     (get-buffer-window
+      (get-file-buffer
+       kbd-mode-demo-file)))))
+
+;;; Helper Functions
+
+(defun kbd-mode--create-demo-buffer ()
+  "Created *kmonad-demo* buffer."
+  (unless (get-buffer "*kmonad-demo*")
+    (display-buffer
+     (get-buffer-create "*kmonad-demo*")
+     `(,(if (version< "24.3" emacs-version)
+            'display-buffer-at-bottom
+          nil)
+       (window-height . 0.15)))))
+
+(defun kbd-mode--find-kbd-file (&optional file)
+  "Choose .kbd file, this file `kbd-mode-demo-file' will used in demo buffer.
+
+If \\[kbd-mode-start-demo] is called from a .kbd file, without SELECTION, it'll be `kbd-mode-demo-file'."
+  (if (and file
+           (string= (file-name-extension file) "kbd"))
+      file
+    (expand-file-name
+     (read-file-name "Choose configuration file"))))
+
+(defun kbd-mode--valid-configp ()
+  "Check if current configuration file is valid."
+  (let ((command (kbd-mode--get-config-validation-command)))
+    (eq 0 (shell-command command))))
+
+(defun kbd-mode--create-kmonad-process ()
+  "Start demo kmonad process."
+  (when (get-buffer-process "*kmonad*")
+    (kbd-mode--kill-kmonad-process))
+  (start-process "kmonad-emacs" "*kmonad*" "kmonad" kbd-mode-demo-file))
+
+(defun kbd-mode--kill-kmonad-process ()
+  "Kill demo kmonad process.
+
+If there is nothing to kill then do nothing."
+  (when (get-buffer-process "*kmonad*")
+    (kill-process "*kmonad*")))
+
+(defun kbd-mode--old-kmonad-suspend ()
+  "Suspend old kmonad instance."
+  (if kbd-mode-old-kmonad-suspend-command
+      (shell-command  kbd-mode-old-kmonad-suspend-command)
+    (error "Set kbd-mode-old-kmonad-suspend-command variable!")))
+
+(defun kbd-mode--old-kmonad-resume ()
+  "Resume old kmonad instance."
+  (if kbd-mode-old-kmonad-suspend-command
+      (shell-command  kbd-mode-old-kmonad-resume-command)
```

Comment:
This doesn't actually work for processes that don't immediately exit, like a simple `kmonad /path/to/config`.  We want to a) call the process asynchronously, so Emacs doesn't block and b) spawn a new process so that, if we kill Emacs, kmonad isn't killed as well

Reply:
My bad, I only tried with `systemd`. I think `async-shell-command` fixes it, but it creates a useless buffer. Manually adding `&` is an easier solution.

Reply:
`async-shell-command` still doesn't fix 

> b) spawn a new process so that, if we kill Emacs, kmonad isn't killed as well

Reply:
Maybe usage of `kill-emacs-hook` can save?

Reply:
you can use `call-process-shell-command`

Reply:
Best way to solve this!